common/src/main/java/net/opentsdb/data/ZonedNanoTimeStamp.java

```diff
@@ -76,53 +76,69 @@
 public class ZonedNanoTimeStamp implements TimeStamp {
   /** The timestamp. */
   private ZonedDateTime timestamp;
+  private volatile long epoch;
+  private volatile long nanos;
   
   public ZonedNanoTimeStamp(final ZonedDateTime timestamp) {
     if (timestamp == null) {
       throw new IllegalArgumentException("Timestamp cannot be null.");
     }
     this.timestamp = ZonedDateTime.from(timestamp);
+    epoch = Instant.from(timestamp).getEpochSecond();
+    nanos = Instant.from(timestamp).getNano();
   }
   
   public ZonedNanoTimeStamp(long epoch_millis, ZoneId zone) {
     if (zone == null) {
       throw new IllegalArgumentException("Zone cannot be null.");
     }
     timestamp = ZonedDateTime.ofInstant(Instant.ofEpochMilli(epoch_millis), zone);
+    epoch = Instant.from(timestamp).getEpochSecond();
```

Comment:
You can also do `millisec/1000` to minimize garbage. `Instant.from(timestamp)` would create a new instance of `Instant` every time. In fact twice for the current code. One for second and one more for nano.

Reply:
Good catch. I still need to use the timesamp functions though as we need to convert from a zoned timestamp to the Unix epochs.